pkg/controller/bar/controller_test.go

```diff
@@ -0,0 +1,141 @@
+	// because NewController returns controller.Interface.
+	controller, ok := NewController(
+		kubeClient,
+		sampleClient,
+		kubeInformer,
+		sampleInformer,
+	).(*Controller)
+	if !ok {
+		t.Fatal("cast to *Controller failed")
+	}
+
+	// Start the informers. This must happen after the call to NewController,
+	// otherwise there are no informers to be started.
+	stopCh = make(chan struct{})
+	kubeInformer.Start(stopCh)
+	sampleInformer.Start(stopCh)
+
+	// Run the controller.
+	go func() {
+		if err := controller.Run(2, stopCh); err != nil {
+			t.Fatalf("Error running controller: %v", err)
+		}
+	}()
+
+	return
+}
+
+// Verify that an event is generated when a Bar is created.
+func TestCreateGeneratesEvent(t *testing.T) {
+	_, sampleClient, controller, _, _, stopCh := newRunningTestController(t)
+	testBar := getTestBar()
+
+	// Create an event watcher on the controller's broadcaster. If an event is
+	// created, this will run and close stopCh, ending the test.
+	controller.broadcaster.StartEventWatcher(func(e *corev1.Event) {
+		assert.Equal(t, "Bar", e.InvolvedObject.Kind)
+		assert.Equal(t, "samplecontroller.k8s.io", e.InvolvedObject.APIVersion)
+		assert.Equal(t, testBar.Name, e.InvolvedObject.Name)
+		assert.Equal(t, MessageResourceSynced, e.Message)
+		assert.Equal(t, SuccessSynced, e.Reason)
+		assert.Equal(t, corev1.EventTypeNormal, e.Type)
+		close(stopCh)
+	})
+
+	// Create a testBar. This should cause an event to be created.
+	sampleClient.SamplecontrollerV1alpha1().Bars("test").Create(testBar)
+
+	// Wait up to 3 seconds for stopCh to close, otherwise fail the test.
+	select {
+	case <-stopCh:
+		return
+	case <-time.After(time.Second * 3):
+		t.Fatal("timed out waiting for event")
+	}
+}
+
+// Verify that an event is generated when a Bar is updated.
+func TestUpdateGeneratesEvent(t *testing.T) {
+	_, sampleClient, controller, _, _, stopCh := newRunningTestController(t)
+
+	testBar := getTestBar()
+	assert.Empty(t, testBar.Spec.DeploymentName)
+
+	// Create an event to update.
+	sampleClient.SamplecontrollerV1alpha1().Bars("test").Create(testBar)
+
+	// Create an event watcher on the controller's broadcaster. If an event is
+	// created, this will run and close stopCh, ending the test.
+	controller.broadcaster.StartEventWatcher(func(e *corev1.Event) {
+		assert.Equal(t, "Bar", e.InvolvedObject.Kind)
+		assert.Equal(t, "samplecontroller.k8s.io", e.InvolvedObject.APIVersion)
+		assert.Equal(t, testBar.Name, e.InvolvedObject.Name)
+		assert.Equal(t, "test-deployment", testBar.Spec.DeploymentName)
+		assert.Equal(t, MessageResourceSynced, e.Message)
+		assert.Equal(t, SuccessSynced, e.Reason)
+		assert.Equal(t, corev1.EventTypeNormal, e.Type)
+		close(stopCh)
+	})
+
+	// Modify the deploymentName of the test Bar
+	testBar.Spec.DeploymentName = "test-deployment"
+
+	// Create a testBar. This should cause an event to be created.
+	sampleClient.SamplecontrollerV1alpha1().Bars("test").Update(testBar)
```

Comment:
@grantr I still don't think this is WAI to drive the controller logic.

If I set up an event watcher like this, and try to get it to `glog.Errorf` both a `Create` and an `Update` (using a channel to ensure the events aren't coalesced), **I only ever see the first event**.

Without the channel logic to try and get multiple events, I see a single event with the changes combined.